libbpf-tools/tcpconnect.bpf.c

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on tcpconnect(8) from BCC by Brendan Gregg
+#include "vmlinux.h"
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#include "kern_map_helpers.h"
+#include "tcpconnect.h"
+
+/*
+ * this hack is needed because otherwise <bpf/bpf_endian.h> includes
+ * <asm-generic/posix_types.h> which conflicts with "vmlinux.h"
+ */
+#define __ASSEMBLY__
+#include <bpf/bpf_endian.h>
+
+const volatile bool do_count;
+const volatile int filter_ports_len;
+SEC(".rodata") int filter_ports_array[MAX_PORTS];
+const volatile pid_t filter_pid;
+const volatile uid_t filter_uid = -1;
+
+/* Define here, because there are conflicts with include files */
+#define AF_INET		2
+#define AF_INET6	10
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, struct sock *);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} sockets SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv4_flow_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} ipv4_count SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct ipv6_flow_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} ipv6_count SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline bool filter_port(__u16 port)
+{
+	if (filter_ports_len == 0)
+		return false;
+
+	for (int i = 0; i < filter_ports_len; i++) {
+		if (port == filter_ports_array[i])
+			return false;
+	}
+	return true;
+}
+
+static __always_inline int
+enter_tcp_connect(struct pt_regs *ctx, struct sock *sk)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 uid = bpf_get_current_uid_gid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = pid_tgid;
+
+	if (filter_pid && pid != filter_pid)
+		return 0;
+
+	uid = bpf_get_current_uid_gid();
+	if (filter_uid != (uid_t) -1 && uid != filter_uid)
+		return 0;
+
+	bpf_map_update_elem(&sockets, &tid, &sk, 0);
+	return 0;
+}
+
+static  __always_inline void count_v4(struct sock *sk, __u16 dport)
+{
+	struct ipv4_flow_key key = {};
+	static __u64 zero;
+	volatile __u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, sk, __sk_common.skc_rcv_saddr);
+	BPF_CORE_READ_INTO(&key.daddr, sk, __sk_common.skc_daddr);
+	key.dport = dport;
+	val = bpf_map_lookup_or_try_init(&ipv4_count, &key, &zero);
+	if (val)
+		*val += 1;
```

Comment:
probably better to use atomic increment here?

Reply:
Yes, thanks, replaced with `__atomic_add_fetch(val, 1, __ATOMIC_RELAXED)`